### ISSUE TITLE
Phantom parameters of values in `where` clauses

### ISSUE THREAD

**Issue by viktorcsimma** [agda/agda2hs#210]
**Update:** it seems to be wrong with functions, too. This test code:

```agda
open import Agda.Builtin.Nat

f1 : Nat -> Nat
f1 n = h1
  where
  g1 : Nat
  g1 = 3 + n
  h1 : Nat
  h1 = n + g1
{-# COMPILE AGDA2HS f1 #-}
```

gets compiled into this:

```hs
module PhantomParameters where

import Numeric.Natural (Natural)

f1 :: Natural -> Natural
f1 n = h1
  where
    g1 :: Natural
    g1 = 3 + n
    h1 :: Natural
    h1 = n + g1 n
```

---

The following test code:

```agda
open import Agda.Builtin.Nat

record Test (a : Set) : Set₁ where
  field
    f : a -> a
open Test {{...}} public
{-# COMPILE AGDA2HS Test class #-}

instance
  testNat : Test Nat
  Test.f testNat n = h
    where
    g : Nat
    g = 3 + n
    h : Nat
    h = n + g
  {-# COMPILE AGDA2HS testNat #-}
```

gets compiled to this Haskell code:

```hs
module PhantomParameters where

import Numeric.Natural (Natural)

class Test a where
    f :: a -> a

instance Test Natural where
    f n = h
      where
        g :: Natural
        g = 3 + n
        h :: Natural
        h = n + g n
```

So `g` gets an extra `n` parameter (probably because it depends on `n`). ~~Interestingly, this problem only occurs when writing instances; in normal functions, this is not a problem.~~

Commit 79fe184a did not contain the bug; so it has to be something that has happened since then.

**Comment by viktorcsimma** [comment]
I think it has to be something [in this commit](https://github.com/agda/agda2hs/pull/205/commits/7d8e44054fd0dd604a2c28595b5ec4ef246489bf). When I reverted that, this problem disappeared; but then there was #185 happening again. However, I don't yet understand what's happening here.

Here is the relevant section of the log at verbosity 20:
```
Compiling module:  PhantomParameters
Compiling definition:  f1
Function locals:  [PhantomParameters.g1, PhantomParameters.h1]
compiling function:  f1
Compiling top-level type Nat → Nat
Compiling type Nat → Nat
Compiling type Nat
Compiling type Nat
applying clauses to parameters:  []
Clause locals:  [PhantomParameters.g1, PhantomParameters.h1]
compiling clause:  <unnamedclause> n = PhantomParameters.h1 n
compiling term: PhantomParameters.h1 n
Compiling application of regular function
Locals:  [PhantomParameters.g1, PhantomParameters.h1]
compiling function:  PhantomParameters.g1
Compiling top-level type (n₁ : Nat) → Nat
Compiling type Nat
applying clauses to parameters:  [n]
Clause locals:  []
compiling clause:  <unnamedclause> = 3 + n
compiling term: 3 + n
Compiling application of regular function
Locals:  []
compiling argument 3
compiling term: 3
compiling argument n
compiling term: n
compiling function:  PhantomParameters.h1
Compiling top-level type (n₁ : Nat) → Nat
Compiling type Nat
applying clauses to parameters:  [n]
Clause locals:  []
compiling clause:  <unnamedclause> = n + PhantomParameters.g1
compiling term: n + PhantomParameters.g1
Compiling application of regular function
Locals:  []
compiling argument n
compiling term: n
compiling argument PhantomParameters.g1
compiling term: PhantomParameters.g1
Compiling application of regular function
Locals:  []
compiling argument n
compiling term: n
Compiling definition:  PhantomParameters.g1
Compiling definition:  PhantomParameters.h1
Writing /home/csimma/Downloads/PhantomParameters.hs
```
So as I can see, it doesn't recognise `n` as a local.

**Comment by viktorcsimma** [comment]
I could identify the specific code block (`Term.hs:241`):

```hs
-            -- Drop module parameters (unless projection-like)
-            n <- (theDef <$> getConstInfo f) >>= \case
-              Function{ funProjection = Right{} } -> return 0
-              _ -> size <$> lookupSection (qnameModule f)
=======
+           reportSDoc "agda2hs.compile.term" 12 $ text "Compiling application of regular function"
+           -- Drop module parameters of local `where` functions
+           xs <- asks locals
+           reportSDoc "agda2hs.compile.term" 15 $ text "Locals: " <+> (prettyTCM xs)
+           n <- if
+             | f `elem` xs -> size <$> lookupSection (qnameModule f)
+             | otherwise   -> return 0
```
When I revert that, this problem is solved and #185 comes back.

If I get it right, the new block leaves too many parameters in the code, while the old one removes too many. But maybe if `asks locals` doesn't contain everything we need, the problem is elsewhere. @jespercockx do you have any ideas?
